packages/vite-plugin-cloudflare/src/index.ts

```diff
@@ -407,67 +410,53 @@ export function cloudflare(pluginConfig: PluginConfig = {}): vite.Plugin[] {
+			configEnvironment() {
 				return {
 					resolve: {
-						alias: getNodeCompatAliases(),
+						builtins: getNodeCompatExternals(),
+					},
+					optimizeDeps: {
+						// We don't want to follow any `node:*` type imports when optimizing deps,
+						// whether or not they are built-in to workerd or polyfilled here.
+						// The ones that are not built-in will just get resolved to their polyfill on demand.
+						exclude: [...nodeBuiltInModules],
+						// But we do want to ensure that all the possible entry-points for the polyfills get pre-bundled
+						// so that they are ready for the first request to the dev server.
+						include: [...getNodeCompatEntries()],
 					},
 				};
 			},
-			configEnvironment(environmentName) {
-				// Ignore Node.js external modules when building environments that use Node.js compat.
-				const workerConfig = getWorkerConfig(environmentName);
-				if (isNodeCompat(workerConfig)) {
-					return {
-						build: {
-							rollupOptions: {
-								external: getNodeCompatExternals(),
-							},
-						},
-					};
-				}
-			},
 			async resolveId(source, importer, options) {
-				// Handle the virtual modules that come from Node.js compat aliases.
-				const strippedSource = maybeStripNodeJsVirtualPrefix(source);
-				if (!strippedSource) {
-					return;
-				}
-
-				const workerConfig = getWorkerConfig(this.environment.name);
-				if (!isNodeCompat(workerConfig)) {
-					// We are not in Node.js compat mode, so we must not try to apply any unenv aliases.
-					// Just resolve the module id as normal.
-					return this.resolve(strippedSource, importer, options);
+				// See if we can map the `source` to a Node.js compat alias.
+				const result = resolveNodeJSImport(source);
+				if (!result) {
+					// The source is not a Node.js compat alias so just pass it through
+					return this.resolve(source, importer, options);
 				}
 
-				// Resolve this id following any unenv aliases.
-				const { unresolved, resolved } = resolveNodeJSImport(strippedSource);
-
 				if (this.environment.mode === "dev" && this.environment.depsOptimizer) {
-					// Make sure the dependency optimizer is aware of this aliased import
-					const optimized =
-						this.environment.depsOptimizer.registerMissingImport(
-							unresolved,
-							resolved
-						).id;
-					// We must pass the id from the depsOptimizer through the rest of the
-					// resolving pipeline to ensure that the optimized version gets used.
-					return this.resolve(optimized, importer, options);
+					// We are in dev mode (rather than build).
+					// Node.js compat polyfill modules have already been pre-bundled,
+					// so we can use the unresolved path to the polyfill
+					// and let the dependency optimizer resolve the path to the cached version.
+					return this.resolve(result.unresolved, importer, options);
 				}
 
-				return this.resolve(resolved, importer, options);
+				// We are in build mode so return the absolute path to the polyfill.
```

Comment:
Nit:
Here you say that we are in build mode, which I think it's correct, but I couldn't tell by the `if` statement above:
```ts
if (this.environment.mode === "dev" && this.environment.depsOptimizer) {
```

could we remove the `&& this.environment.depsOptimizer` and use an `assert` instead?
Something like:
![Screenshot 2025-02-17 at 22 16 00](https://github.com/user-attachments/assets/c4119b99-21ee-446c-a276-72d83af58440)

I think this would make the code easier to follow 🙂
(and having the `depsOptimizer` is a hard requirement in dev for us, isn't it?)